tests/unittests/test_salesforce_date_windowing.py

```diff
@@ -0,0 +1,167 @@
+
+@mock.patch('tap_salesforce.salesforce.Bulk._close_job')
+@mock.patch('tap_salesforce.salesforce.Bulk._poll_on_pk_chunked_batch_status', side_effect = mocked_batch_status)
+@mock.patch('tap_salesforce.salesforce.Bulk._create_job', side_effect=[1, 2, 3, 4, 5])
+@mock.patch('tap_salesforce.salesforce.bulk.singer_utils.now')
+class TestBulkDateWindow(unittest.TestCase):
+
+    # start date
+    start_date = '2019-02-04T12:15:00Z'
+    # 'Salesforce' object
+    sf = Salesforce(
+        default_start_date=start_date,
+        api_type="BULK")
+    # dummy catalog entry
+    catalog_entry = {
+        'stream': 'User',
+        'tap_stream_id': 'User',
+        'schema': {
+            "properties": {
+                "Id": {
+                    "type": "string"
+                },
+                "SystemModstamp": {
+                    "anyOf": [{
+                            "type": "string",
+                            "format": "date-time"
+                        },
+                        {
+                            "type": [
+                                "string",
+                                "null"
+                            ]
+                        }
+                    ]
+                }
+            }
+        },
+        'metadata': [
+            {
+                "breadcrumb": [],
+                "metadata": {
+                    "selected": True,
+                    "replication-key": "SystemModstamp",
+                    "table-key-properties": [
+                        "Id"
+                    ]
+                }
+            },
+            {
+                "breadcrumb": [
+                    "properties",
+                    "SystemModstamp"
+                ],
+                "metadata": {
+                    "inclusion": "automatic"
+                }
+            },
+            {
+                "breadcrumb": [
+                    "properties",
+                    "Id"
+                ],
+                "metadata": {
+                    "inclusion": "automatic"
+                }
+            }
+        ]
+    }
+
+    # mocked now date
+    now_date_1 = datetime.datetime(2022, 5, 2, 12, 15, 00, tzinfo=tz.UTC)
+    now_date_1_str = now_date_1.strftime("%Y-%m-%dT%H:%M:%SZ")
+
+    # mocked now date
+    now_date_2 = datetime.datetime(2019, 2, 5, 12, 15, 00, tzinfo=tz.UTC)
+
+    @mock.patch('tap_salesforce.salesforce.Bulk._add_batch')
+    def test_bulk_date_windowing(self, mocked_add_batch, mocked_singer_util_now, mocked_create_job, mocked_batch_status, mocked_close_job):
+        """
+        To verify that if data is too large then date windowing mechanism execute properly similar to REST api date windowing
+        """
+
+        mocked_singer_util_now.return_value = self.now_date_1
+        Bulk(self.sf)._bulk_with_window([], self.catalog_entry, self.start_date)
+
+        self.assertEqual(mocked_add_batch.call_count, 3, "Function is not called expected times")
```

Comment:
Should we also assert on the dates we expect the 3 calls to happen with?

Reply:
Removed this test case as we are already doing that in the test case:
https://github.com/singer-io/tap-salesforce/blob/6735251097a9c2f79b53ae8a1638b70e7cf1af2e/tests/unittests/test_salesforce_date_windowing.py#L119-L124